// Copyright (C) 2022 Rodrigo Jose Hernandez Cordoba
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";
package AeonGames;

import "vector3.proto";
import "vector4.proto";

message SixDOPMsg
{
    Vector3Msg positive = 1;
    Vector3Msg negative = 2;
}

message BrushMsg {
    SixDOPMsg SixDOP                 = 1;
    uint32  PlaneStart               = 2;
    uint32  PlaneCount               = 3;
}

message KdTreeNodeMsg{
    uint32 Axis                      = 1;
    float Distance                   = 2;
    sint32 Near                      = 3;
    sint32 Far                       = 4;
}

message KdTreeLeafMsg{
    uint32  BrushStart               = 1;
    uint32  BrushCount               = 2;
}

message CollisionMsg {
	uint32 Version                   = 1;
    Vector3Msg Center                = 2;
    Vector3Msg Radii                 = 3;
    repeated sint32     PlaneIndices = 4;
    repeated uint32     BrushIndices = 5;
    repeated Vector4Msg Planes       = 6;
    repeated BrushMsg   Brushes      = 7;
    repeated KdTreeNodeMsg KdNodes   = 8;
    repeated KdTreeLeafMsg KdLeaves  = 9;
}
